reactor-netty5-http/src/main/java/reactor/netty5/http/server/HttpServerFormDecoderProvider.java

```diff
@@ -296,4 +335,120 @@ HttpServerFormDecoderProvider build() {
 		}
 	}
+
+	static final class ReactorNettyHttpPostMultipartRequestDecoder extends HttpPostMultipartRequestDecoder
+			implements ReactorNettyHttpPostRequestDecoder {
+
+		/**
+		 * Current {@link HttpData} from the body (only the completed {@link HttpData}).
+		 */
+		final List<HttpData> currentCompletedHttpData = new ArrayList<>();
+
+		ReactorNettyHttpPostMultipartRequestDecoder(HttpDataFactory factory, HttpRequest request) {
+			super(factory, request);
+		}
+
+		@Override
+		protected void addHttpData(InterfaceHttpData data) {
+			if (data instanceof HttpData httpData) {
+				currentCompletedHttpData.add(httpData);
+			}
+		}
+
+		@Override
+		public void cleanCurrentHttpData(boolean onlyCompleted) {
+			for (HttpData data : currentCompletedHttpData) {
+				removeHttpDataFromClean(data);
+				data.close();
+			}
+			currentCompletedHttpData.clear();
+
+			if (!onlyCompleted) {
+				InterfaceHttpData partial = currentPartialHttpData();
+				if (partial instanceof HttpData partialData) {
+					partialData.delete();
+				}
+			}
+		}
+
+		@Override
+		public List<HttpData> currentHttpData(boolean onlyCompleted) {
+			if (!onlyCompleted) {
+				InterfaceHttpData partial = currentPartialHttpData();
+				if (partial instanceof HttpData partialData) {
+					partialData.usingContent(partialContent ->
+							currentCompletedHttpData.add(partialData.replace(partialContent.split())));
+				}
+			}
+
+			return currentCompletedHttpData;
+		}
+
+		@Override
+		public void destroy() {
+			super.destroy();
+			InterfaceHttpData partial = currentPartialHttpData();
+			if (partial != null && partial.isAccessible()) {
+				partial.close();
+			}
+		}
+	}
+
+	static final class ReactorNettyHttpPostStandardRequestDecoder extends HttpPostStandardRequestDecoder
+			implements ReactorNettyHttpPostRequestDecoder {
+
+		/**
+		 * Current {@link HttpData} from the body (only the completed {@link HttpData}).
+		 */
+		final List<HttpData> currentCompletedHttpData = new ArrayList<>();
+
+		ReactorNettyHttpPostStandardRequestDecoder(HttpDataFactory factory, HttpRequest request) {
+			super(factory, request);
+		}
+
+		@Override
+		protected void addHttpData(InterfaceHttpData data) {
+			if (data instanceof HttpData httpData) {
+				currentCompletedHttpData.add(httpData);
+			}
+		}
+
+		@Override
+		public void cleanCurrentHttpData(boolean onlyCompleted) {
+			for (HttpData data : currentCompletedHttpData) {
+				removeHttpDataFromClean(data);
+				data.close();
+			}
+			currentCompletedHttpData.clear();
+
+			if (!onlyCompleted) {
+				InterfaceHttpData partial = currentPartialHttpData();
+				if (partial instanceof HttpData partialData) {
+					partialData.delete();
+				}
+			}
+		}
+
+		@Override
+		public List<HttpData> currentHttpData(boolean onlyCompleted) {
+			if (!onlyCompleted) {
+				InterfaceHttpData partial = currentPartialHttpData();
+				if (partial instanceof HttpData partialData) {
+					partialData.usingContent(partialContent ->
```

Comment:
Which operation is throwing an exception so that we need to use lambda?

Reply:
I don't understand the difference between `HttpData#copy` and `HttpData#replace`.
and why we cannot invoke directly one of these methods.

Reply:
No operation is throwing anything here. 

The _HttpData.usingContent()_ lambda replaces the old _HttpData.content()_; so to get the content of the http data, you need to pass a lambda which is called with the content. And if the lambda is throwing any checked exceptions, it will be propagated to the caller method which is calling HttpData.usingContent().
The _HttpData.usingContent()_ method uses type inference and allows to propagate any exceptions thrown by the supplied lambda (if it throws any checked exceptions). So, here the lambda does not throw any exceptions.

Some context about why the _HttpData.content() method has been replaced by _HttpData.usingContent()_: the old _HttpData.content()_ ownership was not clear: for disk based http data, a copy of the content was returned and the buffer ownership was returned to the caller, who was assumed to close the buffer. But for memory based http data, the internal buffer was returned, but not the ownership, so no need to close the returned buffer. So now, for disk based http data, the buffer passed to the _usingContent_ lambda method is closed automatically when the lambda returns, so the ownership is not transferred to the caller.

Please check the netty-contrib issue for all the details: https://github.com/netty-contrib/codec-multipart/pull/17, and there are some use cases in the end of the issue.

Reply:
since the _usingContent_ method is not yet finalized in the codec-multipart, I have committed a conservative and safe patch in [868e1bd](https://github.com/reactor/reactor-netty/pull/2327/commits/868e1bdd9cbdd5fe91e4c2896d70d3b055978220), which copies the partial content into the currentCompletedHttpData list. 

this is not optimal, but once the multiplart codec will be finalized, we can then revisit the way partial data are handled.